### PR TITLE
core: terraform.Context API using a functional style

### PR DESCRIPTION
**Pull request by apparentlymart**
This PR has, I'll admit, got a bit out of hand compared to what I was originally setting out to do. I'll talk more about the motivating goals later on in this comment, but since the PR is now dominated by a rather large side-adventure I think it better to talk about _that_ first.

---

Previously `terraform.Context` was built in an unfortunate way where all of the data was provided up front in `terraform.NewContext` and then mutated quietly and directly by subsequent operations. That made the data flow hard to follow, commonly leading to bugs, and also meant that we were forced to take various actions too early in `terraform.NewContext`, rather than waiting until a more appropriate time during an operation.

The biggest commit in the middle of this PR changes `terraform.Context` so that its fields are broadly just unchanging data about the execution context (current workspace name, available plugins, etc) whereas the main data Terraform works with arrives via individual method arguments and is returned in return values.

Specifically, this means that `terraform.Context` no longer "has-a" config, state, and "planned changes", instead holding on to those only temporarily during an operation. The caller is responsible for propagating the outcome of one step into the next step so that the data flow between operations is actually visible.

Here's roughly how things would flow before (Go-like pseudocode):

```go
c := terraform.NewContext(&terraform.ContextOpts{
    // (plugins, current workspace, and other bits that remain constant throughout a run)

    Config: config,
    State: prevRunState,
    Variables: inputVariablesExplicitlySet,
    PlanMode: plans.NormalMode,
    // (and targets, forced-replace, and various other operation-specific thingies)
})

diags := c.Input()
// (deal with diags)
diags = c.Validate()
// (deal with diags)
_, diags := c.Plan()
// (deal with diags)
_, diags = c.Apply()
// (deal with diags)
newState := c.State()
// and then write newState as a new snapshot to persistent state storage
```

Here's the equivalent of that pseudocode with the new API:

```go
c := terraform.NewContext(&terraform.ContextOpts{
    // (plugins, current workspace, and other bits that remain constant throughout a run)

})

diags := c.Input(config)
// (deal with diags)
diags = c.Validate(config)
// (deal with diags)
plan, diags := c.Plan(config, prevRunState, &terraform.PlanOpts{
    Mode: plans.NormalMode,
    Config: config,
    InputState: prevRunState,
    SetVariables: inputVariablesExplicitlySet,
    // (and targets, forced-replace, and various other plan-specific thingies)
})
// (deal with diags)
newState, diags = c.Apply(plan, config)
// (deal with diags)
// and then write newState as a new snapshot to persistent state storage
```

It's admittedly pretty subtle in terms of just raw diff, but this means that `terraform.NewContext` is no longer a huge function responsible for processing everything before we even start an operation, and we can more clearly see which inputs each step takes and how the outputs from one step flow into later steps. Something not visible from the pseudocode alone is that these functions now avoid mutating the objects given as input, so e.g. `prevRunState` in the above remains distinct from `newState` and doesn't automatically reflect any of the changes made by `c.Plan` or `c.Apply`.

However, since that's a change to the main entry points in the `terraform` package, this commit also touches every file in the codebase which interacted with those APIs. Most of the noise here is in updating tests to take the same actions using the new API style, but this also affects the main-code callers in the backends and in the `command` package.

My goal here was to refactor without changing externally-observable behavior, but in practice there are a couple externally-visible behavior variations here that seemed okay in service of the broader goal, vs. the opportunity cost of working to preserve them exactly:

- The "terraform graph" command is no longer hooked directly into the core graph builders, because that's no longer part of the public API. A secondary goal of my design decisions here was to make graph building and walking just an implementation detail of Terraform Core's behavior, rather than it being so explicit in the exported API.

    However, I did include a couple new `Context` functions whose contract is to produce a _UI-oriented graph_, and _for now_ those continue to return the physical graph we use for the plan and apply operations. There's no exported API for generating the "validate" and "eval" graphs, because neither is particularly interesting in its own right, and so `terraform graph` no longer supports those graph types.

    We intentionally excluded `terraform graph` behavior from the Terraform v1.0 Compatibility Promises because we knew it was previously exposing implementation details and those details were likely to change in future. Therefore although is is technically a breaking change, it's one that ought to have minimal-to-low impact, and is within the bounds of what we previously promised.

- `terraform.NewContext` no longer has the responsibility for collecting all of the provider schemas up front. Instead, we wait until we need them.

    However, that means that some of our error messages now have a slightly different shape due to unwinding through a differently-shaped call stack.

    As of this commit we also end up reloading the schemas multiple times in some cases, which is functionally acceptable but likely represents a performance regression due to the overhead of spinning up and calling into plugins each time. I intend to rework this to use caching per provider address, but I'm saving that for a later commit because this one is big enough already and that'll be only a performance optimization, not a significant behavior change.

The proximal reason for this change is to resolve the chicken/egg problem whereby there was previously no single point where we could apply "moved" statements to the previous run state before creating a plan, which I'll discuss more below.

However, this is at least the third project in a row where the previous API design led to piling more functionality into `terraform.NewContext` and then working around the incorrect order of operations that produces, so I intend that by paying the cost/risk of this large diff now we can in turn reduce the cost/risk of future projects that relate to our main workflow actions.

Although this _largely_ tidies up the `terraform.Context` API to have a more visible/appropriate data flow, there are a couple remaining flaws I left unattended for now for future work:

- The `Context.Input` method still modifies a hidden internal map of provider settings to record the result of the input prompts. As the pre-existing commentary in that method continues to say, it's pretty weird that the `terraform` package gets involved in interactive UI prompts at all, but we still have some remaining architectural quirks blocking us from fully fixing that.
- Although the mental model for a plan includes both prior state _and_ the input configuration, the way we model that in saved plan files (as a "config snapshot" rather than as a `*configs.Config`) means that we can't actually match that logical model in the physical data structures. Instead, `Context.Apply` expects the caller to provide a configuration equivalent to the one that was previously passed to `Context.Plan`, and would misbehave if a caller provided a modified configuration by mistake.
- We record checksums of the provider plugins that contributed to a plan as a part of that plan, but management of plugins is one of the responsibilities that remains with the `Context` object as a whole rather than the `Apply` method in particular, and so a caller trying to apply a saved plan needs to pass part of the data from the `*plans.Plan` object into the `terraform.NewContext` call, even though the _rest_ of the plan object is handled exclusively by `Context.Apply`.
- The existing backend API has its own share of oddities relating to over-generalization of the various operations it supports, but we haven't yet completed a detailed design for a replacement mechanism for switching between local and remote operations, and so unfortunately I had to introduce a new concept up at the backend layer (`backend.LocalRun`) which aims to preserve the "big bag of stuff" model that `terraform.Context` used to follow, so that refactoring of the backend layer can wait for another day. The new design should be functionally equivalent to the old, but it is admittedly unfortunate to make that layer slightly worse (by introducing a new concept) before we can make it better. Still, on the plus side I think this limits a lot of our historical architecture problems now _just_ to the backend layer, which we are planning to revisit and redesign in the future anyway.
- (Some other minor bits and bobs I've noted in comments inline, but won't repeat here because this comment is already more than long enough.)

Since this was gonna be a pretty gnarly diff anyway, I also took the opportunity to fix some latent `go vet` and `staticcheck` lints we've accumulated over the years -- particularly in our old tests -- under our usual "clean up only while you're in the area anyway" principle.

---

The original goal here had been to connect together the remaining dots to make Terraform handle `moved` blocks and then report any changes that resulted as part of the plan, and so this PR _also_ largely achieves that, though I stopped short once I realized that the PR was going to end up including the big meaty refactor commit.

Specifically, for config-based move this achieves:

- There's a place in the plan model to remember the "previous run address" for each resource instance mentioned in the plan, which will differ from the current address if we're processing a move as part of this plan.
- The main resource planning operation now checks the "move result" and sets the previous run address appropriately based on what it found there.
- The `Context.Plan(...)` operation applies any pending moves to the previous run state _before_ diverging into the separate previous run state, refreshed state, and working state artifacts, whereas before it was happening too late and thus applying the moves _only_ to the previous run state, which is largely the same as not applying the moves at all.

I've split the parts that are for the config-based move project into separate commits, in the hope of making it easier to review that functionality in smaller doses rather than it getting swamped by the noisy refactoring commit in the middle. Ideally I would've split this into two separate PRs, but the refactoring was invasive enough that it wasn't really practical to extricate it from work I'd already done prior to realizing that the refactoring was necessary. Sorry for such a noisy, mixed-issue PR; pragmatism must prevail here, sadly. :confounded: 

The main remaining steps related to end-to-end config-based move _after_ this PR would be:
- Make the UI layer (in the `command` and `backend/local` packages) aware of the concept of a "previous run address" for a resource instance and present that in a reasonable way both in the human-oriented and in the machine-oriented (JSON) plan descriptions.
- Fix the logic which populates the `refactoring.MoveResult` map while processing moves so that it'll produce correct descriptions of the result of processing chained and nested moves. Currently it produces a corrupted data structure in some cases, because it only partially updates a pre-existing result when executing another move that follows on from it. (There's a comment in the `refactoring` package with more details on this.)
- Complete the validation rules for `moved` blocks, which currently only handles some simpler cases. In particular, I've preserved a pre-existing TODO here for detecting the situation where a necessary move has been excluded from consideration by `-target`, which we need to detect and handle because otherwise it could make the resulting state and plan inconsistent.

### CODE REVIEW

**Review by jbardin — APPROVED** [review]
Thanks for the _huge_ context refactor too, rather than packing another mutable option in somewhere!

**Review by alisdair — APPROVED** [review]
Looks great! My eyes started to glaze over with the context test fixups, but I think I understood everything else. This is much clearer, and thanks for taking the time to split out the commits 🙌

**Review by chrisarcand — APPROVED** [review]
Extremely impressive amount of context recording in this PR, nice work!

**Comment by kmoe** [comment]
This PR changes the error text on no init from "Error: could not load plugin" to "Please run \"terraform init\"". Will this change land in a v1.0.x version or v1.1.0?

**Comment by jbardin** [comment]
Hi @kmoe, this change is currently only in `main`, and would be slated for the v1.1 release.
